httprule/client.go

```diff
@@ -0,0 +1,118 @@
+package httprule
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	pb "google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	descpb "google.golang.org/protobuf/types/descriptorpb"
+)
+
+type ClientConn struct {
+	HTTPClient *http.Client
+	BaseURL    string
+}
+
+var (
+	ErrInvalidMethod    = errors.New("invalid gRPC method string")
+	ErrMethodNotFound   = errors.New("method not found")
+	ErrServiceNotFound  = errors.New("service not found")
+	ErrNotImplemented   = errors.New("not implemented")
+	ErrHttpRuleNotFound = errors.New("no HttpRule")
+)
+
+func (c *ClientConn) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return nil, ErrNotImplemented
+}
+
+func (c *ClientConn) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
+	rule, err := getHttpRule(method)
+	if err != nil {
+		return err
+	}
+	req, err := NewHTTPRequest(rule, c.BaseURL, args.(proto.Message))
+	if err != nil {
+		return err
+	}
+	resp, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode > 399 {
+		return errorStatus(resp.StatusCode)
+	}
+	return ParseProtoResponse(rule, resp.Body, reply.(proto.Message))
+}
+
+func getHttpRule(method string) (*pb.HttpRule, error) {
+	parts := strings.Split(method, "/")
+	if len(parts) != 3 || parts[0] != "" {
+		return nil, fmt.Errorf("%w: %s", ErrInvalidMethod, method)
+	}
+
+	serviceName, methodName := protoreflect.FullName(parts[1]), protoreflect.Name(parts[2])
+	desc, err := protoregistry.GlobalFiles.FindDescriptorByName(serviceName)
+	if err != nil {
+		return nil, fmt.Errorf("%w, %v", ErrServiceNotFound, err)
+	}
+
+	sd, ok := desc.(protoreflect.ServiceDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("service is not a service: %s", serviceName)
```

Comment:
this error msg reads a little funny. perhaps
```
return nil, fmt.Errorf("%w: '%s' is not a service", ErrInvalidMethod, serviceName)
```

I'm going with "invalid" but it could also be ErrServiceNotFound too

Reply:
I kinda did that on purpose.

I took not from your method descriptor is not a method descripor.

it is pretty much impossible to simulte a test for this. I'm almost tempted call this an internal error or panic on it...

I've changed this now - but it was on purpose. maybe you have any other ideas about what to do with unreachable errors?

Reply:
Couldn't you test with a method path of `/HelloRequest/Hello` ?